pkg/apis/resource/validation/validation.go

```diff
@@ -621,27 +666,75 @@ func validateResourcePool(pool resource.ResourcePool, fldPath *field.Path) field
+	allErrs = append(allErrs, validateMap(device.Capacity, -1, attributeAndCapacityMaxKeyLength, validateQualifiedName, validateDeviceCapacity, fldPath.Child("capacity"))...)
 	if combinedLen, max := len(device.Attributes)+len(device.Capacity), resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice; combinedLen > max {
 		allErrs = append(allErrs, field.Invalid(fldPath, combinedLen, fmt.Sprintf("the total number of attributes and capacities must not exceed %d", max)))
 	}
+
+	allErrs = append(allErrs, validateSet(device.ConsumesCapacity, resource.ResourceSliceMaxDeviceCapacityConsumptions,
+		func(deviceCapacityConsumption resource.DeviceCapacityConsumption, fldPath *field.Path) field.ErrorList {
+			return validateDeviceCapacityConsumption(deviceCapacityConsumption, fldPath, capacityPoolNames)
+		},
+		func(deviceCapacityConsumption resource.DeviceCapacityConsumption) (string, string) {
+			return deviceCapacityConsumption.CapacityPool, "capacityPool"
+		}, fldPath.Child("consumesCapacity"))...)
+
+	if perDeviceNodeSelection {
+		numDeviceNodeSelectionFields := 0
+		if device.NodeName != "" {
+			numDeviceNodeSelectionFields++
+			allErrs = append(allErrs, validateNodeName(device.NodeName, fldPath.Child("nodeName"))...)
+		}
+		if device.NodeSelector != nil {
+			numDeviceNodeSelectionFields++
+			allErrs = append(allErrs, corevalidation.ValidateNodeSelector(device.NodeSelector, false, fldPath.Child("nodeSelector"))...)
+		}
+		if device.AllNodes {
+			numDeviceNodeSelectionFields++
+		}
+		switch numDeviceNodeSelectionFields {
+		case 0:
+			allErrs = append(allErrs, field.Required(fldPath, "exactly one of `nodeName`, `nodeSelector`, or `allNodes` is required when `perDeviceNodeSelection` is set in the ResourceSlice spec"))
+		case 1:
+		default:
+			allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one of `nodeName`, `nodeSelector`, or `allNodes` is required when `perDeviceNodeSelection` is set in the ResourceSlice spec"))
+		}
+	} else if device.NodeName != "" || device.NodeSelector != nil || device.AllNodes {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "`nodeName`, `nodeSelector` and `allNodes` can only be set if `perDeviceNodeSelection` is set in the ResourceSlice spec"))
+	}
+
+	return allErrs
+}
+
+func validateDeviceCapacityConsumption(deviceCapacityConsumption resource.DeviceCapacityConsumption, fldPath *field.Path, capacityPoolNames sets.Set[string]) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if !capacityPoolNames.Has(deviceCapacityConsumption.CapacityPool) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("capacityPool"), deviceCapacityConsumption.CapacityPool, "must reference a capacity pool defined in the ResourceSlice"))
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.DRAPartitionableDevices) && deviceCapacityConsumption.Capacity == nil {
```

Comment:
Rather than checking for nil, prefer to use `len()` - it works on nil and empty, in case of weirdness in decoding